analysis/pseudobulk-bulk-prediction/exploratory-notebooks/pseudobulk-calculations.Rmd

```diff
@@ -0,0 +1,455 @@
+make_binned_scatterplots(
+  project_df_list$SCPCP000009, 
+  project_id = "SCPCP000009",
+  nbins = 15,
+  facet_rows = 1
+)
+```
+
+
+```{r fig.height = 8, fig.width = 12, message=FALSE, warning=FALSE}
+make_binned_scatterplots(
+  project_df_list$SCPCP000017, 
+  project_id = "SCPCP000017",
+  nbins = 40,
+  facet_rows = 7
+)
+```
+
+* `bulk_counts` generally has much closer to a 1:1 relationship with pseudobulk compared to `bulk_tpm`
+  * This is nearly universally the case for `SCPCP000001`, `SCPCP000002`, and `SCPCP000006`, and while true for `SCPCP000009` and `SCPCP000017`, it is less pronounced
+* Relationships in general are weakest for `SCPCP000017`, but the situation does look better with `bulk_counts`
+
+
+#### Statistics
+
+Let's nail this down further with correlations comparing each measure for bulk with each measure for pseudobulk.
+We'll perform correlations on a per-sample basis, both parametric and non-parametric, display some correlations below both as boxplots and the full table.
+
+```{r fig.height = 10, fig.width = 6}
+
+# Helper function to run correlations
+model_samples <- function(id, df) {
+  sample_df <- df |>
+    dplyr::filter(sample_id == id) 
+  
+  #### Bulk tpm
+  df_deseq <- sample_df |>
+    dplyr::filter(is.finite(pseudobulk_deseq), is.finite(bulk_tpm))
+  r_deseq_tpm <- cor(df_deseq$bulk_tpm, df_deseq$pseudobulk_deseq, method = "spearman")
+  rho_deseq_tpm <- cor(df_deseq$bulk_tpm, df_deseq$pseudobulk_deseq, method = "pearson")
+
+  df_log_counts <- sample_df |>
+      dplyr::filter(is.finite(pseudobulk_log_counts), is.finite(bulk_tpm))
+  r_logcounts_tpm <- cor(df_deseq$bulk_tpm, df_deseq$pseudobulk_log_counts, method = "spearman")
+  rho_logcounts_tpm <- cor(df_deseq$bulk_tpm, df_deseq$pseudobulk_log_counts, method = "pearson")
+      
+  #### Bulk counts
+  df_deseq <- sample_df |>
+    dplyr::filter(is.finite(pseudobulk_deseq), is.finite(bulk_counts))
+  r_deseq_counts <- cor(df_deseq$bulk_counts, df_deseq$pseudobulk_deseq, method = "spearman")
+  rho_deseq_counts <- cor(df_deseq$bulk_counts, df_deseq$pseudobulk_deseq, method = "pearson")
+
+  df_log_counts <- sample_df |>
+      dplyr::filter(is.finite(pseudobulk_log_counts), is.finite(bulk_counts))
+  r_logcounts_counts <- cor(df_deseq$bulk_counts, df_deseq$pseudobulk_log_counts, method = "spearman")
+  rho_logcounts_counts <- cor(df_deseq$bulk_counts, df_deseq$pseudobulk_log_counts, method = "pearson")
+      
+  # Tabulate and return some fit stats
+  data.frame(
+    r =      c(r_deseq_tpm,    r_logcounts_tpm,    r_deseq_counts,   r_logcounts_counts),
+    rho =    c(rho_deseq_tpm,  rho_logcounts_tpm,  rho_deseq_counts, rho_logcounts_counts),
+    bulk =   c("bulk_tpm",    "bulk_tpm",          "bulk_counts",    "bulk_counts"), 
+    pseudo = c("deseq",       "log_counts",        "deseq",          "log_counts")
+  ) |>
+    tidyr::pivot_longer(
+      c(r, rho), 
+      names_to = "measure", 
+      values_to = "correlation"
+    )
+}
```

Comment:
This seems be a good case for `tidyr::expand_grid()` and `purrr::pmap()`, but honestly, it's fine. Especially for this.

Reply:
At one point there was a comment in here that I realize got deleted along the way, but it said something like "this is not the most efficient way to do this, but it gets the job done"